src/openpgp/openpgp.cpp

```diff
@@ -0,0 +1,380 @@
+  }
+
+  const auto hash_algorithm = buffer.read_big_endian<uint8_t>();
+
+  const auto hashed_data_length = buffer.read_big_endian<uint16_t>();
+  std::vector<uint8_t> hashed_data = buffer.read(hashed_data_length);
+
+  const auto unhashed_data_length = buffer.read_big_endian<uint16_t>();
+  buffer.read_span(unhashed_data_length);
+
+  const uint16_t hash_left16_bits = *reinterpret_cast<const uint16_t *>(&buffer.read_span(sizeof(uint16_t))[0]);
+
+  const mpi signature = buffer.read_mpi();
+
+  return signature_rsa(
+    algorithm,
+    hash_left16_bits,
+    hash_algorithm,
+    hashed_data,
+    signature_type,
+    s_expression("(sig-val (rsa (s %m)))", signature.get()),
+    version);
+}
+
+bool signature_rsa::verify(const epee::span<const uint8_t> message, const public_key_rsa &public_key) const
+{
+  const s_expression signed_data = hash_message(message, public_key.bits());
+  return gcry_pk_verify(m_signature.get(), signed_data.get(), public_key.get()) == 0;
+}
+
+s_expression signature_rsa::hash_message(const epee::span<const uint8_t> message, size_t public_key_bits) const
+{
+  switch (m_type)
+  {
+  case type::binary_document:
+    return hash_bytes(message, public_key_bits);
+  case type::canonical_text_document:
+  {
+    std::vector<uint8_t> crlf_formatted;
+    crlf_formatted.reserve(message.size());
+    const size_t message_size = message.size();
+    for (size_t offset = 0; offset < message_size; ++offset)
+    {
+      const auto &character = message[offset];
+      if (character == '\r')
+      {
+        continue;
+      }
+      if (character == '\n')
+      {
+        const bool skip_last_crlf = offset + 1 == message_size;
+        if (skip_last_crlf)
+        {
+          break;
+        }
+        crlf_formatted.push_back('\r');
+      }
+      crlf_formatted.push_back(character);
+    }
+    return hash_bytes(epee::to_span(crlf_formatted), public_key_bits);
+  }
+  default:
+    throw std::runtime_error("unsupported signature type");
+  }
+}
+
+std::vector<uint8_t> signature_rsa::hash_asn_object_id() const
+{
+  size_t size;
+  if (gcry_md_algo_info(m_hash_algorithm, GCRYCTL_GET_ASNOID, nullptr, &size) != GPG_ERR_NO_ERROR)
+  {
+    throw std::runtime_error("failed to get ASN.1 Object Identifier (OID) size");
+  }
+
+  std::vector<uint8_t> asn_object_id(size);
+  if (gcry_md_algo_info(m_hash_algorithm, GCRYCTL_GET_ASNOID, &asn_object_id[0], &size) != GPG_ERR_NO_ERROR)
+  {
+    throw std::runtime_error("failed to get ASN.1 Object Identifier (OID)");
+  }
+
+  return asn_object_id;
+}
+
+s_expression signature_rsa::hash_bytes(const epee::span<const uint8_t> message, size_t public_key_bits) const
+{
+  const std::vector<uint8_t> plain_hash = (hash(m_hash_algorithm) << message << m_hashed_appendix).finish();
+  if (plain_hash.size() < sizeof(m_hash_left16_bits))
+  {
+    throw std::runtime_error("insufficient message hash size");
+  }
+  if (*reinterpret_cast<const uint16_t *>(&plain_hash[0]) != m_hash_left16_bits)
```

Comment:
That looks endian dependent (might not care ?)

Reply:
Update. Is a pair of bytes now.

Reply:
Sorry to complain again, but AFAIK function parameter evaluation order is undefined (and that's a function in disguise).